lib/collection/src/collection_manager/holders/segment_holder.rs

```diff
@@ -761,25 +776,77 @@ impl<'s> SegmentHolder {
         Ok((proxies, tmp_segment, segments_lock))
     }
 
+    /// Try to unproxy a single shard segment for [`proxy_all_segments_and_apply`]
+    ///
+    /// # Warning
+    ///
+    /// If unproxying fails an error is returned with the lock and the proxy is left behind in the
+    /// shard holder.
+    fn try_unproxy_segment(
+        segments_lock: RwLockUpgradableReadGuard<SegmentHolder>,
+        proxy_id: SegmentId,
+        proxy_segment: LockedSegment,
+    ) -> Result<RwLockUpgradableReadGuard<SegmentHolder>, RwLockUpgradableReadGuard<SegmentHolder>>
+    {
+        // We must propagate all changes in the proxy into their wrapped segments, as we'll put the
+        // wrapped segment back into the segment holder. This can be an expensive step if we
+        // collected a lot of changes in the proxy, so we do this in two batches to prevent
+        // unnecessary locking. First we propagate all changes with a read lock on the shard
+        // holder, to prevent blocking other readers. Second we propagate any new changes again
+        // with a write lock on the segment holder, blocking other operations. This second batch
+        // should be very fast, as we already propagated all changes in the first, which is why we
+        // can hold a write lock. Once done, we can swap out the proxy for the wrapped shard.
+
+        let proxy_segment = match proxy_segment {
+            LockedSegment::Proxy(proxy_segment) => proxy_segment,
+            LockedSegment::Original(_) => {
+                log::warn!("Unproxying segment {proxy_id} that is not proxified, that is unexpected, skipping");
+                return Err(segments_lock);
+            }
+        };
+
+        // Batch 1: propagate changes to wrapped segment with segment holder read lock
+        if let Err(err) = proxy_segment.read().propagate_to_wrapped() {
+            log::error!("Propagating proxy segment {proxy_id} changes to wrapped segment failed, ignoring: {err}");
+        }
+
+        let mut write_segments = RwLockUpgradableReadGuard::upgrade(segments_lock);
+
+        // Batch 2: propagate changes to wrapped segment with segment holder write lock
+        // Propagate proxied changes to wrapped segment, take it out and swap with proxy
+        let wrapped_segment = {
+            let proxy_segment = proxy_segment.read();
+            if let Err(err) = proxy_segment.propagate_to_wrapped() {
+                log::error!("Propagating proxy segment {proxy_id} changes to wrapped segment failed, ignoring: {err}");
+            }
+            proxy_segment.wrapped_segment.clone()
+        };
+        let (_, segments) = write_segments.swap(wrapped_segment, &[proxy_id]);
```

Comment:
Hmm, can do, let me think about it for a bit.

The interesting thing about this method is that it only _tries_ to unproxy the segment. It can safely fail or postpone it. For example, if we don't have a strong reference count of 1.

If it does not unproxy here, it'll unproxy one more time at the end of `proxy_all_segments_and_apply()`.

Reply:
I just remembered we have `unwrap_with_timeout` somewhere that we use for this purpose, _kinda_ feels it might be worth considering here as well?

Reply:
Yes, we use that when calling `drop_data` on a segment. If the temporary segment gets no data at all, it is called on that.

Maybe we do need to get an exclusive Arc though, because I don't think we want to fill the `deleted_points`, `deleted_indexes` and `created_indexes` after we have unproxified at the segment holder level. Wdyt?

Reply:
I take a write lock on the segment holder in the second batch. The idea was that it'd ensure exclusive access. But that isn't the case of course since we can clone the proxy segment Arc.

Reply:
Yeah, it kinda seems like a rather narrow corner case, like, it's probably gonna be very hard to trigger in production. But it seems like it might be a potential race-condition/concurrency bug. But probably can ignore it for now.